### PR TITLE
Add domain policy

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/aspnet/Routing/issues/825

Note that this is based on https://github.com/aspnet/Routing/pull/799

### CODE REVIEW

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            // are then then execute them in linear order.
+            var ordered = edges
+                .Select(e => (domain: (HostString)e.State, destination: e.Destination))
+                .OrderBy(e => GetScore(e.domain))
+                .ToArray();
+
+            return new DomainPolicyJumpTable(exitDestination, ordered);
+        }
+
+        private int GetScore(in HostString domain)
```
Comment (rynowak) [inline samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: This probably deserves a comment about why it's not used by the comparer. It makes sense to me. But I might forget.

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+        public IReadOnlyList<PolicyNodeEdge> GetEdges(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            // The algorithm here is designed to be preserve the order of the endpoints
+            // while also being relatively simple. Preserving order is important.
```
Comment (rynowak) [inline samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: This comment is designed to be totally original, while also being a copy paste from something I wrote. S P I C Y

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                    {
+                        kvp.Value.Add(endpoint);
+                    }
+                }
+                else
+                {
+                    // OK this endpoint matches specific domains
+                    foreach (var kvp in edges)
+                    {
+                        // The edgeKey maps to a possible request header value
```
Comment (rynowak) [inline samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: host?

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                        var edgeKey = kvp.Key;
+
+                        for (var j = 0; j < domains.Count; j++)
+                        {
+                            var domain = domains[j];
+
+                            if (edgeKey == domain)
```
Comment (rynowak) [inline samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: Is this logic correct when you have wildcards? It sounds like it's not, but you expect the ordering of the domains to help you. Consider a case like this:

Endpoint1:
domain: foo.example.com
version: v2

Endpoint2:
domain *.example.com
version: v1

This code will be build edges like this:

foo.example.com -> (Endpoint1)
*.example.com -> (Endpoint2)

So what if the request is for foo.example.com - version 1?

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs (hunk 1 above)
Comment (rynowak) [reply samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: Maybe this should be used by the comparer. See https://github.com/aspnet/Routing/pull/847/files#r224504940

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                _destinations = destinations;
+            }
+
+            public override int GetDestination(HttpContext httpContext)
+            {
+                var destinations = _destinations;
+                for (var i = 0; i < destinations.Length; i++)
+                {
+                    if ((string.IsNullOrEmpty(destinations[i].domain.Host) ||
+                        httpContext.Request.Host.Host == destinations[i].domain.Host) &&
```
Comment (rynowak) [inline samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: maybe I misunderstand what you're saying about wildcards? I kinda expected to see handling for something like `*.example.com`

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs (hunk 4 above)
Comment (rynowak) [reply samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: OR 

> maybe I misunderstand what you're saying about wildcards? I kinda expected to see handling for something like *.example.com

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs (hunk 2 above)
Comment (JamesNK) [reply samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: 😋

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs (hunk 5 above)
Comment (JamesNK) [reply samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: By wildcards I meant that it supports *:port, or host:*. I didn't consider wildcards inside the host. That would be a required feature. Will add!

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs (hunk 1 above)
Comment (JamesNK) [reply samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: I copied this from the ConsumesMatcherPolicy. Should it be using the score in the comparer as well?

File: samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs (hunk 3 above)
Comment (JamesNK) [reply samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs]: host is a request header value

Technically correct: The best kind of correct

**Comment by natemcmaster** [comment]
We're merging Routing today into aspnet/AspNetCore. If you can merge this in the next 30 minutes or so, we can bring this along. Otherwise, you'll have to open a new PR against https://github.com/aspnet/AspNetCore.

**Pull request closed without merging**